packages/node-sdk/db/src/pg-connection.ts

```diff
@@ -1,13 +1,86 @@
 import type { Client, PoolClient, PoolConfig } from "pg";
 import pg from "pg";
 import { ComponentNames, log, SeverityNumber } from "@paima/log";
+import { type Operation, run, sleep } from "npm:effection@3.5.0";
+import { ENV } from "@paima/utils";
 
 let readonlyDBConn: pg.Pool | null;
 
-export const getConnection = (creds: PoolConfig, readonly = false): pg.Pool => {
+// PGLite does not support multiple connections, so we need to use a mutex to ensure that only one query is executed at a time.
+// * For transactions use yield* aquireDBMutex(); ...Your Operations... releaseDBMutex();
+// * For single queries use await runPreparedQuery(myQuery.run(params, dbConn));
+// IMPORTANT: This is only for PGLite instances, for full pgsql servers, this is not needed.
+const IS_PGLITE = true; // TODO: make this configurable
+
+// TODO This is a very simple mutex implementation.
+//      It releases the mutex randomly, and not in order.
+//      It only works for deno's single-threaded runtime.
+//      pglite exports async-mutex, so we can use that for a proper solution.
+let db_mutex: "free" | "locked" = "free";
+/**
+ * Acquires a mutex to ensure that only one query is executed at a time.
+ * If running in PGLite, it will acquire a mutex to ensure that only one query is executed at a time.
+ * If running in a full pgsql server, it will do nothing.
+ * Always run run `releaseDBMutex()` after the query locks no longer needed, other threads are blocked until the mutex is released.
+ */
+export function* aquireDBMutex(): Operation<void> {
```

Comment:
see `packages/paima-sdk/utils/src/concurrency` for things you might be able to reuse, or be inspired by

Reply:
PAI-52